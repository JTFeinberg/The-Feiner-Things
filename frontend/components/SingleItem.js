import React, { Component } from 'react'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import styled from 'styled-components'
import Head from 'next/head'
import Error from './ErrorMessage'


const SingleItemStyles = styled.div`
max-width: 1200px;
margin: 2rem auto;
box-shadow: ${props => props.theme.bs};
display: grid;
grid-auto-columns: 1fr;
grid-auto-flow: column;
min-height: 800px;
`

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: {id: $id}){
        id
        title
        description
        largeImage
    }
}
`

export default class SingleItem extends Component {
    render() {
        return (
            <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
                {({ error, loading, data }) => {
                    const { item } = data
                    if (error) return <Error error={error} />
                    if (loading) return <p>Loading...</p>
                    if (!item) return <p>No Item Found for ID {this.props.id}</p>
                    return (
                        <SingleItemStyles>
                            <Head>
                                <title>Feiner Things | {item.title}</title>
                            </Head>
                            <img src={item.largeImage} alt={item.title} />
                            <div className="details">
                                <h2>Viewing {item.title}</h2>
                                <p>{item.description}</p>
                            </div>
                        </SingleItemStyles>
                    )
                }}
            </Query>
        )
    }
}
