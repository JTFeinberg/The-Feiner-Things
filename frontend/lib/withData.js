import withApollo from 'next-with-apollo'
import ApolloClient from 'apollo-boost'
import { endpoint, prodEndpoint } from '../config'
import { LOCAL_STATE_QUERY } from '../components/Cart'

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      })
    },
    // local data
    clientState: {
      resolvers: {
        Mutation: {
          // First argument is unused (ever?) so we  use _
          // Third argument is client, but we dont need the whole thing so we just destructure the cache
          toggleCart(_, variables, { cache }) {
            // Read the cartOpen value from cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY
            })
            // Write the cart state to the opposite
            const data = {
              data: {
                cartOpen: !cartOpen
              }
            }
            cache.writeData(data)
            return data
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  })
}

export default withApollo(createClient)
