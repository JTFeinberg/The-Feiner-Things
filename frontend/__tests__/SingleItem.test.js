import React from 'react'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem'
import { fakeItem } from '../lib/testUtils'

describe('<SingleItem />', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        //when somebody makes a request with this query and variable combo
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        //return this fake data (mocked data)
        result: {
          data: {
            item: fakeItem()
          }
        }
      }
    ]
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    )
  })
})
