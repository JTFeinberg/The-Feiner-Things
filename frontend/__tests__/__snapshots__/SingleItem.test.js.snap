// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;
